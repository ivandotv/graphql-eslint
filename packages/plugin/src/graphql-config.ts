import { GraphQLConfig, GraphQLExtensionDeclaration, loadConfigSync, SchemaPointer } from 'graphql-config';
import { CodeFileLoader } from '@graphql-tools/code-file-loader';
import { ParserOptions } from './types';

let graphQLConfig: GraphQLConfig;

export function loadGraphQLConfig(options: ParserOptions): GraphQLConfig {
  // We don't want cache config on test environment
  // Otherwise schema and documents will be same for all tests
  if (process.env.NODE_ENV !== 'test' && graphQLConfig) {
    return graphQLConfig;
  }

  const onDiskConfig = options.skipGraphQLConfig
    ? null
    : loadConfigSync({
        throwOnEmpty: false,
        throwOnMissing: false,
        extensions: [addCodeFileLoaderExtension],
      });

  const configOptions = options.projects
    ? { projects: options.projects }
    : {
        schema: (options.schema || '') as SchemaPointer, // if `schema` is `undefined` will throw error `Project 'default' not found`
        documents: options.documents || options.operations,
        extensions: options.extensions,
        include: options.include,
        exclude: options.exclude,
      };

  graphQLConfig =
    onDiskConfig ||
    new GraphQLConfig(
      {
        config: configOptions,
        filepath: 'virtual-config',
      },
      [addCodeFileLoaderExtension]
    );

  return graphQLConfig;
}

const addCodeFileLoaderExtension: GraphQLExtensionDeclaration = api => {
  api.loaders.schema.register(new CodeFileLoader());
  api.loaders.documents.register(new CodeFileLoader());
  return { name: 'graphql-eslint-loaders' };
};
