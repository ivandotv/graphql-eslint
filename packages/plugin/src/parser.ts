import { parseGraphQLSDL } from '@graphql-tools/utils';
import { ASTNode, GraphQLError, TypeInfo, Source } from 'graphql';
import { Linter } from 'eslint';
import { convertToESTree } from './estree-parser';
import { GraphQLESLintParseResult, ParserOptions, ParserServices } from './types';
import { extractTokens } from './utils';
import { getSchema } from './schema';
import { getSiblingOperations } from './sibling-operations';
import { loadGraphQLConfig } from './graphql-config';
import { getReachableTypes, getUsedFields } from './graphql-ast';

export function parse(code: string, options?: ParserOptions): Linter.ESLintParseResult['ast'] {
  return parseForESLint(code, options).ast;
}

export function parseForESLint(code: string, options: ParserOptions = {}): GraphQLESLintParseResult {
  const gqlConfig = loadGraphQLConfig(options);
  let schema
  try {
    schema = getSchema(options, gqlConfig);
  } catch (e) {
    e.message = `[graphql-eslint] Error while loading schema: ${e.message}`;
    // eslint-disable-next-line no-console
    console.error(e)
  }

  const parserServices: ParserServices = {
    hasTypeInfo: Boolean(schema),
    schema,
    siblingOperations: getSiblingOperations(options, gqlConfig),
    reachableTypes: getReachableTypes,
    usedFields: getUsedFields,
  };

  try {
    const filePath = options.filePath || '';

    const graphqlAst = parseGraphQLSDL(filePath, code, {
      ...options.graphQLParserOptions,
      noLocation: false,
    });

    const { rootTree, comments } = convertToESTree(graphqlAst.document as ASTNode, schema ? new TypeInfo(schema) : null);
    const tokens = extractTokens(new Source(code, filePath));

    return {
      services: parserServices,
      parserServices,
      ast: {
        type: 'Program',
        body: [rootTree as any],
        sourceType: 'script',
        comments,
        loc: rootTree.loc,
        range: rootTree.range as [number, number],
        tokens,
      },
    };
  } catch (e) {
    e.message = `[graphql-eslint] ${e.message}`;
    // In case of GraphQL parser error, we report it to ESLint as a parser error that matches the requirements
    // of ESLint. This will make sure to display it correctly in IDEs and lint results.
    if (e instanceof GraphQLError) {
      const eslintError = {
        index: e.positions[0],
        lineNumber: e.locations[0].line,
        column: e.locations[0].column,
        message: e.message,
      };
      throw eslintError;
    }
    throw e;
  }
}
