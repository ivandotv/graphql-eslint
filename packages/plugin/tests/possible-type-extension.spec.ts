import { join } from 'path';
import { GraphQLRuleTester, rules } from '../src';

const ruleTester = new GraphQLRuleTester();

ruleTester.runGraphQLTests('possible-type-extension', rules['possible-type-extension'], {
  valid: [
    {
      code: 'extend type User { name: String }',
      parserOptions: {
        schema: 'type User { id: ID }',
      },
    },
    {
      name: 'when schema is separate into graphql files',
      filename: join(__dirname, 'mocks/possible-type-extension/separate-graphql-files/extend-type-user.gql'),
      code: ruleTester.fromMockFile('possible-type-extension/separate-graphql-files/extend-type-user.gql'),
      parserOptions: {
        schema: join(__dirname, 'mocks/possible-type-extension/separate-graphql-files/*.gql'),
      },
    },
    {
      name: 'when schema is separate into code files',
      filename: join(__dirname, 'mocks/possible-type-extension/separate-code-files/extend-type-user.ts'),
      code: /* GraphQL */ `
        extend type User {
          firstName: String
        }
      `,
      parserOptions: {
        schema: join(__dirname, 'mocks/possible-type-extension/separate-code-files/*.ts'),
      },
    },
    {
      name: 'when schema exist in one file',
      filename: join(__dirname, 'mocks/possible-type-extension/one-graphql-file/type-user.gql'),
      code: ruleTester.fromMockFile('possible-type-extension/one-graphql-file/type-user.gql'),
      parserOptions: {
        schema: join(__dirname, 'mocks/possible-type-extension/one-graphql-file/type-user.gql'),
      },
    },
  ],
  invalid: [
    {
      code: 'extend type OtherUser { name: String }',
      parserOptions: {
        schema: 'type User { id: ID }',
      },
      errors: [{ message: 'Cannot extend type "OtherUser" because it is not defined.' }],
    },
  ],
});
