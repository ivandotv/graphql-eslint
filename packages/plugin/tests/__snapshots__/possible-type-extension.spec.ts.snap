// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
> 1 | extend type OtherUser { name: String }
    |             ^^^^^^^^^ Cannot extend type "OtherUser" because it is not defined.
`;
